File: app/src/main/java/com/sample/android/newsreader/app/data/FetchTask.java
package com.sample.android.newsreader.app.data;

import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;

public abstract class FetchTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {

    public interface Callback<Result> {
        void onPreExecute();
        void onPostExecute(@Nullable Result result);
    }

    public interface OnStatusListener {
        void onComplete();
        void onError(@Nullable String message, @Nullable Throwable exception);
    }

    protected Callback<Result> mCallback = null;

    protected OnStatusListener mOnStatusListener = null;

    private boolean mIsTaskLive = false;

    public void setCallback(@Nullable Callback<Result> callback) {
        mCallback = callback;
    }

    public void setOnStatusListener(@Nullable OnStatusListener listener) {
        mOnStatusListener = listener;
    }

    @Override
    protected void onPreExecute() {
        mIsTaskLive = true;
        if (mCallback != null) {
            mCallback.onPreExecute();
        }
        Log.d("FetchTask", "onPreExecute flag=" + mIsTaskLive);
    }

    @SafeVarargs
    @Nullable
    @Override
    protected final Result doInBackground(@Nullable Params... params) {
        try {
            if (params != null) {
                return onBackground(params);
            }
        } catch (Exception e) {
            onError(e);
        }
        return null;
    }

    @Nullable
    @SuppressWarnings("unchecked")
    protected abstract Result onBackground(@NonNull Params... params) throws Exception;

    protected void onError(Exception e) {
        if (mIsTaskLive && mOnStatusListener != null) {
            Log.d("FetchTask", "onError");
            mOnStatusListener.onError(e.getMessage(), e.getCause());
            mIsTaskLive = false;
        }
    }

    @Override
    protected void onPostExecute(@Nullable Result result) {
        if (mIsTaskLive && mOnStatusListener != null) {
            Log.d("FetchTask", "onComplete");
            mOnStatusListener.onComplete();
            mIsTaskLive = false;
        }
        if (mCallback != null) {
            mCallback.onPostExecute(result);
        }
    }
}

File: app/src/main/kotlin/com/sample/android/newsreader/app/data/FetchTask.kt
package com.sample.android.newsreader.app.data

import android.os.AsyncTask
import android.util.Log

abstract class FetchTask<Params, Progress, Result> : AsyncTask<Params, Progress, Result>() {

    interface Callback<Result> {
        fun onPreExecute()
        fun onPostExecute(result: Result)
    }

    interface OnStatusListener {
        fun onComplete()
        fun onError(message: String?, exception: Throwable?)
    }

    protected var mCallback: Callback<Result>? = null

    protected var mOnStatusListener: OnStatusListener? = null

    private var mIsTaskLive = false

    fun setCallback(callback: Callback<Result>?) {
        mCallback = callback
    }

    fun setOnStatusListener(listener: OnStatusListener?) {
        mOnStatusListener = listener
    }

    override fun onPreExecute() {
        mIsTaskLive = true
        if (mCallback != null) {
            mCallback!!.onPreExecute()
        }
        Log.d("FetchTask", "onPreExecute flag=" + mIsTaskLive)
    }

    override fun doInBackground(vararg params: Params): Result? {
        try {
            return onBackground(*params)
        } catch (e: Exception) {
            onError(e)
            return null
        }

    }

    @Throws(Exception::class)
    protected abstract fun onBackground(vararg params: Params): Result?

    protected fun onError(e: Exception) {
        if (mIsTaskLive && mOnStatusListener != null) {
            Log.d("FetchTask", "onError")
            mOnStatusListener!!.onError(e.message, e.cause)
            mIsTaskLive = false
        }
    }

    override fun onPostExecute(result: Result) {
        if (mIsTaskLive && mOnStatusListener != null) {
            Log.d("FetchTask", "onComplete")
            mOnStatusListener!!.onComplete()
            mIsTaskLive = false
        }
        if (mCallback != null) {
            mCallback!!.onPostExecute(result)
        }
    }
}